plugins/hls-explicit-record-fields-plugin/src/Ide/Plugin/ExplicitFields.hs

```diff
@@ -0,0 +1,266 @@
+    LogShake shakeLog -> pretty shakeLog
+
+descriptor :: Recorder (WithPriority Log) -> PluginId -> PluginDescriptor IdeState
+descriptor recorder plId = (defaultPluginDescriptor plId)
+  { pluginHandlers = mkPluginHandler STextDocumentCodeAction codeActionProvider
+  , pluginRules = collectRecordsRule recorder
+  }
+
+codeActionProvider :: PluginMethodHandler IdeState 'TextDocumentCodeAction
+codeActionProvider ideState pId (CodeActionParams _ _ docId range _) = pluginResponse $ do
+  nfp <- getNormalizedFilePath (docId ^. L.uri)
+  pragma <- getFirstPragma pId ideState nfp
+  CRR renderedRecs (map unExt -> exts) <- collectRecordsInRange range ideState nfp
+  let actions = map (mkCodeAction nfp exts pragma) renderedRecs
+  pure $ List actions
+
+  where
+    mkCodeAction :: NormalizedFilePath -> [Extension] -> NextPragmaInfo -> RenderedRecordInfo -> Command |? CodeAction
+    mkCodeAction nfp exts pragma rec = InR CodeAction
+      { _title = mkCodeActionTitle exts
+      , _kind = Just CodeActionRefactorRewrite
+      , _diagnostics = Nothing
+      , _isPreferred = Nothing
+      , _disabled = Nothing
+      , _edit = Just $ mkWorkspaceEdit nfp edits
+      , _command = Nothing
+      , _xdata = Nothing
+      }
+      where
+        edits = catMaybes [ mkTextEdit rec , pragmaEdit ]
+
+        mkTextEdit :: RenderedRecordInfo -> Maybe TextEdit
+        mkTextEdit (RenderedRecordInfo ss r) = TextEdit <$> srcSpanToRange ss <*> pure r
+
+        -- NOTE(ozkutuk): `RecordPuns` extension is renamed to `NamedFieldPuns`
+        -- in GHC 9.4, but we still want to insert `NamedFieldPuns` in pre-9.4
+        -- GHC as well, hence the replacement.
+        -- https://gitlab.haskell.org/ghc/ghc/-/merge_requests/6156
+        pragmaEdit :: Maybe TextEdit
+        pragmaEdit = if NamedFieldPuns `elem` exts
+                       then Nothing
+                       else Just $ patchExtName $ insertNewPragma pragma NamedFieldPuns
+          where
+            patchExtName = L.newText %~ T.replace "Record" "NamedField"
+
+    mkWorkspaceEdit :: NormalizedFilePath -> [TextEdit] -> WorkspaceEdit
+    mkWorkspaceEdit nfp edits = WorkspaceEdit changes Nothing Nothing
+      where
+        changes = Just $ HashMap.singleton (fromNormalizedUri (normalizedFilePathToUri nfp)) (List edits)
+
+    mkCodeActionTitle :: [Extension] -> Text
+    mkCodeActionTitle exts =
+      if NamedFieldPuns `elem` exts
+        then title
+        else title <> " (needs extension: NamedFieldPuns)"
+        where
+          title = "Expand record wildcard"
+
+collectRecordsRule :: Recorder (WithPriority Log) -> Rules ()
+collectRecordsRule recorder = define (cmapWithPrio LogShake recorder) $ \CollectRecords nfp -> do
+  tmr <- use TypeCheck nfp
+  let exts = getEnabledExtensions <$> tmr
+      recs = getRecords <$> tmr
+      renderedRecs = mapMaybe renderRecordInfo <$> recs
+  pure ([], CRR <$> renderedRecs <*> exts)
+
+getEnabledExtensions :: TcModuleResult -> [GhcExtension]
```

Comment:
Hmm, I also wonder if this function is not already written somewhere?

Reply:
I don't think there is one, but this code is duplicated among several plugins so I think it a good idea to write one. Where would be a good place to add this definition?

Reply:
Not sure. We don't have a great home for it at the minute, admittedly. There might be somewhere in ghcide.

Reply:
I have moved it to `ghcide/src/Development/IDE/GHC/Util.hs`, and replaced ad-hoc definitions among different plugins to use this one.